[Dynamo] No graph break for super(MyConv{1/2/3}d, self).forward and super(MyConvTranspose, self).forward

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102509

before the PR, running super(MyConv1d, self).forward or super(MyConvTranspose, self).foward, dynamo will create a graph break when executing NNModuleVariable.call_method and raise unimplemented error for name=_conv_forward / _output_padding. see issue for full detail: https://github.com/pytorch/pytorch/issues/101155

after the PR, for torch.nn.conv module with function name _conv_forward / _output_padding, we inline the function with tx.inline_user_function_return

code refactor: added NNModuleVariable._inline_user_function_return_helper to consolidaste tx.inline_user_function_return into 1 place to keep code dry. after factor, there are 2 uncolidated inline_user_function_return with different ```fn``` and ```source``` logic. the code is still dry. For local testing, they are covered by test_modulelist, test_moduledict, test_conv_call_super_forward_directly and test_conv_transpose_call_super_forward_directly in test_modules.py

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

Differential Revision: [D46494460](https://our.internmc.facebook.com/intern/diff/D46494460)

weifengpy committed:
[Dynamo] No graph break for super(MyConv{1/2/3}d, self).forward

before the PR, running super(MyConv1d, self).forward, dynamo will create a graph break when executing NNModuleVariable.call_method and raise unimplemented error for name=_conv_forward. see issue for full detail: https://github.com/pytorch/pytorch/issues/101155

after the PR, for torch.nn.conv module with function name _conv_forward, we inline the function with tx.inline_user_function_return

[ghstack-poisoned]

willfengg commented:
/easycla

yanboliang commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -606,6 +606,18 @@ def gen_source(source, name):
+            module_attr = getattr(fn, "__module__", "")
+            if module_attr is not None and module_attr.startswith(
+                "torch.nn.modules.conv"
+            ):
```
I would suggest to use ```isinstance(mod, torch.nn.modules.conv._ConvNd```, and you can get ```mod``` by ```tx.output.get_submodule(self.module_key)```. If users extends from the builtin ```Conv``` module, they can benefit from this inline. We do see similar patterns from torchbench models.

yanboliang commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -606,6 +606,18 @@ def gen_source(source, name):
+            if module_attr is not None and module_attr.startswith(
+                "torch.nn.modules.conv"
+            ):
+                return tx.inline_user_function_return(
```
There are many places finally navigate to ```inline_user_function_return```, can you try to consolidate them?

yanboliang reviewed (commented):
This looks correct overall, I left two small comments. BTW, can you also add supports/tests for ```_ConvTransposeNd. _output_padding```? Thank you!

weifengpy committed:
Update on "[Dynamo] No graph break for super(MyConv{1/2/3}d, self).forward"


before the PR, running super(MyConv1d, self).forward, dynamo will create a graph break when executing NNModuleVariable.call_method and raise unimplemented error for name=_conv_forward. see issue for full detail: https://github.com/pytorch/pytorch/issues/101155

after the PR, for torch.nn.conv module with function name _conv_forward, we inline the function with tx.inline_user_function_return

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -599,12 +639,44 @@ def gen_source(source, name):
+                args,
                 kwargs,
+                options,
+                function_type=InlineFunctionType.METHOD,
```
I believe we can consolidate all of these into:
```
elif name == "_get_abs_string_index" or (
    isinstance(module, torch.nn.modules.conv._ConvNd)
    and name == "_conv_forward"
) or (
    isinstance(module, torch.nn.modules.conv._ConvTransposeNd)
    and name == "_output_padding"
):
    # Inline the function
    fn = getattr(module, name).__func__
    fn_source = AttrSource(self.source, "__func__")
    options["source"] = fn_source
    return tx.inline_user_function_return(
        variables.UserFunctionVariable(fn, **options),
        [self] + args,
        kwargs,
    )
```

yanboliang replied:
We can use either ```UserFunctionVariable``` or ```UserMethodVariable``` here, just need to manipulate args to what they expect and update source correctly.

yanboliang replied:
In the future, we can use this function as the default handling logic for ```NNModuleVariable.call_method```.

weifengpy committed:
Update on "[Dynamo] No graph break for super(MyConv{1/2/3}d, self).forward and super(MyConvTranspose, self).forward"


before the PR, running super(MyConv1d, self).forward or super(MyConvTranspose, self).foward, dynamo will create a graph break when executing NNModuleVariable.call_method and raise unimplemented error for name=_conv_forward / _output_padding. see issue for full detail: https://github.com/pytorch/pytorch/issues/101155

after the PR, for torch.nn.conv module with function name _conv_forward / _output_padding, we inline the function with tx.inline_user_function_return

code refactor: added NNModuleVariable._inline_user_function_return_helper to consolidaste tx.inline_user_function_return into 1 place to keep code dry. after factor, there are 2 uncolidated inline_user_function_return with different ```fn``` and ```source``` logic. the code is still dry. For local testing, they are covered by test_modulelist, test_moduledict, test_conv_call_super_forward_directly and test_conv_transpose_call_super_forward_directly in test_modules.py

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

weifengpy committed:
Update on "[Dynamo] No graph break for super(MyConv{1/2/3}d, self).forward and super(MyConvTranspose, self).forward"


before the PR, running super(MyConv1d, self).forward or super(MyConvTranspose, self).foward, dynamo will create a graph break when executing NNModuleVariable.call_method and raise unimplemented error for name=_conv_forward / _output_padding. see issue for full detail: https://github.com/pytorch/pytorch/issues/101155

after the PR, for torch.nn.conv module with function name _conv_forward / _output_padding, we inline the function with tx.inline_user_function_return

code refactor: added NNModuleVariable._inline_user_function_return_helper to consolidaste tx.inline_user_function_return into 1 place to keep code dry. after factor, there are 2 uncolidated inline_user_function_return with different ```fn``` and ```source``` logic. the code is still dry. For local testing, they are covered by test_modulelist, test_moduledict, test_conv_call_super_forward_directly and test_conv_transpose_call_super_forward_directly in test_modules.py

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

weifengpy committed:
Update on "[Dynamo] No graph break for super(MyConv{1/2/3}d, self).forward and super(MyConvTranspose, self).forward"


before the PR, running super(MyConv1d, self).forward or super(MyConvTranspose, self).foward, dynamo will create a graph break when executing NNModuleVariable.call_method and raise unimplemented error for name=_conv_forward / _output_padding. see issue for full detail: https://github.com/pytorch/pytorch/issues/101155

after the PR, for torch.nn.conv module with function name _conv_forward / _output_padding, we inline the function with tx.inline_user_function_return

code refactor: added NNModuleVariable._inline_user_function_return_helper to consolidaste tx.inline_user_function_return into 1 place to keep code dry. after factor, there are 2 uncolidated inline_user_function_return with different ```fn``` and ```source``` logic. the code is still dry. For local testing, they are covered by test_modulelist, test_moduledict, test_conv_call_super_forward_directly and test_conv_transpose_call_super_forward_directly in test_modules.py

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

yanboliang commented:
@pytorchbot  merge

willfengg commented:
@willfengg has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46494460).
